Support large files in multipart for resteasy reactive

Sgitario opened the pull request:
I've added tests for resteasy reactive and also resteasy reactive rest client. 

Fix https://github.com/quarkusio/quarkus/issues/24472

michalszynkiewicz commented on extensions/resteasy-reactive/quarkus-resteasy-reactive/deployment/src/test/java/io/quarkus/resteasy/reactive/server/test/multipart/MultipartOutputResource.java:
```diff
@@ -19,6 +21,7 @@
     public static final List<String> RESPONSE_VALUES = List.of("one", "two");
     public static final boolean RESPONSE_ACTIVE = true;
 
+    private static final int ONE_GIGA = 1024 * 1024 * 1024;
```
the exception that I (just) pasted in the issue said specifically that at ca 2*10^9, we should test sth like 3G

Sgitario replied:
I could reproduce it with 1GB only in my machine. I tried with 3gb and worked fine too. However, I think using 1gb will be better in terms of CI resources (if we keep this test)

michalszynkiewicz commented on extensions/resteasy-reactive/quarkus-resteasy-reactive/deployment/src/test/java/io/quarkus/resteasy/reactive/server/test/multipart/MultipartOutputUsingBlockingEndpointsTest.java:
```diff
@@ -18,7 +18,7 @@
 
     private static final String EXPECTED_CONTENT_DISPOSITION_PART = "Content-Disposition: form-data; name=\"%s\"";
-    private static final String EXPECTED_CONTENT_DISPOSITION_FILE_PART = "Content-Disposition: form-data; name=\"%s\"; filename=\"%s\"";
+    private static final String EXPECTED_CONTENT_DISPOSITION_FILE_PART = "Content-Disposition: form-data; name=\"%s\"; filename=\"%s";
```
why did this change?

michalszynkiewicz commented on extensions/resteasy-reactive/rest-client-reactive/deployment/src/test/java/io/quarkus/rest/client/reactive/multipart/MultipartResponseTest.java:
```diff
@@ -209,15 +223,24 @@ public MultipartData getFile() throws IOException {
+        public MultipartData getLargeFile() throws IOException {
+            File file = createTemporalFileToDownload();
+            RandomAccessFile f = new RandomAccessFile(file, "rw");
+            f.setLength(ONE_GIGA);
```
I think we should also have verification if the file content is okay

michalszynkiewicz replied:
I don't see it (other than file existence)

Sgitario replied:
I've just updated the PR now.

gsmet commented:
I don’t mind adding the test but I think we should condition their executions to a system property so that they are not run by default.

michalszynkiewicz commented on independent-projects/resteasy-reactive/client/runtime/src/main/java/org/jboss/resteasy/reactive/client/impl/ClientSerialisers.java:
```diff
@@ -104,10 +103,10 @@ public static Buffer invokeClientWriter(Entity<?> entity, Object entityObject, C
-                        entity.getMediaType(), headerMap, baos);
-                return Buffer.buffer(baos.toByteArray());
+                        entity.getMediaType(), headerMap, out);
+                return out.getBuffer();
```
I'm probably missing a lot, but isn't it still keeping it all in a single buffer before sending?

Sgitario replied:
Before these changes, we were storing the large file in a byte array (ByteArrayOutputStream). 
With these changes, we will directly use the buffer which works similarly than BufferedOutputStream.

geoand replied:
What Michal is asking IIUC, is whether all the data is still in memory or not.

geoand replied:
Let me try and rephrase: If one is to send a 20GB file back, how much data is in memory during the sending operation? It is a constant number, or does it increase with file size?

Asking because, if the number is not constant, then sure, this fix is better than what we had, but it's not **the** best fix (for example, in the other two similar issues I fixed yesterday, I had to drop down to Vert.x handling of the body because Vert.x already handles everything properly).

geoand replied:
Great, thanks

Sgitario commented:
@geoand and @gsmet , about disabling this test, if we keep using 1gb, from my point of view is not that large, so it should be acceptable for the CI (unless the CI results start complaining about it of course).

geoand commented:
I am not sure it is... Best have a system property like @gsmet mentions

Sgitario commented:
> I am not sure it is... Best have a system property like @gsmet mentions

@gsmet is there a system property so this test is disabled by default, but not on a daily basis? My whole point is to ensure we don't break this functionality in the future.

gsmet commented:
@Sgitario we could enable it in the early access nightly build for instance.

geoand commented:
That's a great idea.

I'll then do that for the fixes I made yesterday that didn't have tests

famod commented:
It's worth a try, but FTR I'd like to mention that general heap pressure is pretty high in that job because it's the only one that's basically running (almost) everything in one mvn call.

michalszynkiewicz commented:
> It's worth a try, but FTR I'd like to mention that general heap pressure is pretty high in that job because it's the only one that's basically running (almost) everything in one mvn call.

If this works properly, it should not consume a lot of heap

Sgitario committed:
Support large files in multipart for resteasy reactive

I've added tests for resteasy reactive and also resteasy reactive rest client. 

Fix https://github.com/quarkusio/quarkus/issues/24472

Sgitario commented:
> @Sgitario we could enable it in the early access nightly build for instance.
> 
> Something like `-Dtest-resteasy-reactive-large-files`.

@michalszynkiewicz I've just updated the PR, but keeping 1gb to test large files. I could reproduce it with this size locally.

@gsmet let me know if I did correctly what you meant to enable these tests only in the early access nightly build. I did test that it's skipping the test by default and including it when adding `-Dtest-resteasy-reactive-large-files` in the Maven command.

michalszynkiewicz commented:
@geoand knows the server part much better than me but this looks good to me (I just expected some pump will be needed, hence the concerns)
